### PR TITLE
Allow binary truncation length to be zero to handle evaluators that encounter empty string values

### PR DESCRIPTION
**Pull request by kbendick**
This PR closes https://github.com/apache/iceberg/issues/2065

This allows for the `BinaryUtil.truncateBinary` function to allow for a zero length truncation. This is necessary for handling
scans where STARTS_WITH filters are applied to the table (or any filter that uses  `truncateBinary`) when a (possibly non-partition) column has an empty string value in it.

It would be more efficient to return early in each of the given metrics evaluators if the length is zero, but I did not want to introduce too many changes that conflict with this PR: https://github.com/apache/iceberg/pull/2062. If we'd like, we can go in and update all of those functions individually, but this is the minimum possible change set (to the non-test code) to fix the issue now and is likely cleaner. Additionally, since we're returning the same empty final ByteBuffer right away, there's not really that much overhead, and certainly no more than currently exists.

I chose to update the `TestFilteredScan` spark test suites to have one record be an empty string. This will cover unpartitioned tables, tables with filters that are partitioned on a column that has the empty string, as well as tables that are partitioned on columns that do not have the empty string value (which was the case for the original ticket).

I also originally added a test that uses the Iceberg API directly in `BaseTableTest`, but it was not a partitioned table as `writeRecords` in that test suite assumes an unpartitioned table spec. I decided to remove it as those code paths are all handled in `TestFilteredScan` without adding any additional overhead to the CI suite. I did however add a test that better documents the (new) behavior of `truncateBinary`.

cc @changquanyou @aokolnychyi @RussellSpitzer

### CODE REVIEW

File: data/src/test/java/org/apache/iceberg/data/TestLocalScan.java [hunk 1]
```diff
@@ -498,6 +499,35 @@ public void testFilterWithDateAndTimestamp() throws IOException {
...
+
+    DataFile file = writeFile(tableLocation.toString(), format.addExtension("record-file"),
+            SCHEMA, testRecords);
+    table.newFastAppend().appendFile(file).commit();
+    Iterable<Record> results = IcebergGenerics.read(table).where(startsWith("data", "")).build();
+    Set<Record> actual = Sets.newHashSet(results);
+
+    Assert.assertEquals("Should produce correct number of records", expected.size(), actual.size());
+    Assert.assertEquals("Record set should match", expected, actual);
+  }
```
Comment (kbendick) [inline data/src/test/java/org/apache/iceberg/data/TestLocalScan.java:529]: I'm thinking of removing this test and just updating the records in the two spark `TestFilteredScan` suite, as that covers a range of possible scenarios (including the one that the original issue was written for) and would not add any additional CI time.

File: api/src/main/java/org/apache/iceberg/util/BinaryUtil.java [hunk 2]
```diff
@@ -28,12 +28,23 @@
...
 
+  private static final ByteBuffer EMPTY_BYTE_BUFFER = ByteBuffer.allocate(0);
+
   /**
-   * Truncates the input byte buffer to the given length
+   * Truncates the input byte buffer to the given length.
+   * <p>
+   * We allow for a length of zero so that rows with empty string can be evaluated.
+   * Partition specs still cannot be created with a length of zero due to a constraint
+   * when parsing column truncation specs in org.apache.iceberg.MetricsModes.
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/util/BinaryUtil.java:38]: I tried using an `{@link}` here but the java docs couldn't find that class because it's in a different module.

File: data/src/test/java/org/apache/iceberg/data/TestLocalScan.java (hunk 1 above)
Comment (kbendick) [reply data/src/test/java/org/apache/iceberg/data/TestLocalScan.java]: I decided to do this instead.

File: api/src/main/java/org/apache/iceberg/util/BinaryUtil.java [hunk 3]
```diff
@@ -28,12 +28,23 @@
   private BinaryUtil() {
   }
 
+  private static final ByteBuffer EMPTY_BYTE_BUFFER = ByteBuffer.allocate(0);
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/util/BinaryUtil.java:31]: I can make this lazily instantiated if we care. Or just return a new empty byte buffer instance every time in the function, but seems unnecessary to have all of those allocations.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java [hunk 4]
```diff
@@ -529,7 +529,7 @@ private Table buildPartitionedTable(String desc, PartitionSpec spec, String udf,
     return Lists.newArrayList(
         record(schema, 0L, parse("2017-12-22T09:20:44.294658+00:00"), "junction"),
         record(schema, 1L, parse("2017-12-22T07:15:34.582910+00:00"), "alligator"),
-        record(schema, 2L, parse("2017-12-22T06:02:09.243857+00:00"), "forrest"),
+        record(schema, 2L, parse("2017-12-22T06:02:09.243857+00:00"), ""),
```
Comment (kbendick) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: I can make a separate test if we feel that would be better. However, this test suite covers a very large number of cases so it seems like a smart place to add it.

**Review by jackye1995 — COMMENTED** [review]
This looks fine to me, given the tests all passed. But I wonder why empty string was not handled and added for testing from the very beginning, I may lack some context here. I see @aokolnychyi added the check for `length > 0`, was there any particular reason for that? Do you expect `length == 0` to be handled before the pushdown?

File: api/src/main/java/org/apache/iceberg/util/BinaryUtil.java (hunk 2 above)
Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/util/BinaryUtil.java:38]: Pushing a version using @code. Thanks for the tip.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java (hunk 4 above)
Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: I also think that it's a good opportunity to start introducing more edge case data into the tests without adding any overhead, and it tests this code path in a myriad of ways and does not affect the original test (though I'll have to align my open PR that currently touches this file, but that's not a problem at all).

**Comment by kbendick** [comment]
> This looks fine to me, given the tests all passed. But I wonder why empty string was not handled and added for testing from the very beginning, I may lack some context here. I see @aokolnychyi added the check for `length > 0`, was there any particular reason for that? Do you expect `length == 0` to be handled before the pushdown?

Yes I would also be very interested to hear if there's any reason to not allow for the truncation function to alter the precondition. I'm open to there being something I didn't catch.

For reference, I checked and the only place that `truncateBinary` is used is in metrics evaluation - namely for `startsWith` (and soon to be notStartsWith).

To be exact, in `InclusiveMetricsEvaluator`,  `ManifestEvaluator`, and `ParquetMetricsRowGroupFilter` (all of them in the visitor and exclusively with startsWith - which is where the bug occurs presumably due to a row with an empty string). It's also used in updating the lower bounds and upper bounds for min/max parquet metrics for `FIXED` and `BINARY` types (in `ParquetUtils`). So it's only used with metrics and literals and it's overall relatively well tested. I can add more tests as a follow up on the binary and fixed types if we feel it necessary, but overall those are all decently well covered in the tests - especially after I add my tests from the `NOT_STARTS_WITH` PR, which covers it further: https://github.com/apache/iceberg/pull/2062.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java (hunk 4 above)
Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: Also, for reference, I did make several separate tests and they all passed. But it just seems cleaner to me to introduce more common edge cases into the existing tests cases. I'm surprised that the tests that run on random data hasn't caught this before. But the tests in `TestFilteredScan` all go over the code paths that deal with this. And after we merge this, I can add in two small / inexpensive tests in the `NOT_STARTS_WITH` pr that will more explicitly cover this edge case - by partitioning a DF using a filter of `data like '%'` and `data not like '%'`, which previously threw an exception due to the truncation length until this change is added.

Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: So I actually found an inexpensive existing place to test `truncateBinary` itself. So the function is not better documented by means of the tests (if we agree that this change on the preconditions is acceptable).

**Comment by kbendick** [comment]
> This looks fine to me, given the tests all passed. But I wonder why empty string was not handled and added for testing from the very beginning, I may lack some context here. I see @aokolnychyi added the check for `length > 0`, was there any particular reason for that? Do you expect `length == 0` to be handled before the pushdown?

As for whether or not length == 0 should be handled before the pushdown, it should be if I understand correctly. By the time that this function is invoked in the places I listed elsewhere, all expressions should have been rewritten to remove NOT nodes, and then been bound. Given that this is always called either in metrics evaluation or row value evaluation against a predicate - START_WITH or NOT_STARTS_WITH, the expressions being evaluated are all `BoundLiteralPredicates` and the input of an empty predicate in that situation (e.g. `data not like '%'`) is not necessarily an invalid expression, though it does get rewritten to be `data != ''` via expression parsing, binding, and rewriting. I'm not as closely familiar with the `min` and `max` functions, but my understanding is that they too should be under these same conditions and only applied to parquet metadata after expressions have also become rewritten `BoundLiteralPredicate`s.

Please correct me if I am mistaken 🙂 .

**Comment by kbendick** [comment]
It seems that CI failed due possibly to timeouts (unable to find something in a catalogue in an unrelated test). I'm rerunning locally, but I think it's just a timed out test that failed due to transient issues.

**Comment by kbendick** [comment]
I'm going to open a PR that adds a check for length 0 when configuring the metrics mode to be Truncate. That's likely why this was originally not allowed to be zero (to disallow having a column with metrics `truncate[0]`). However, this isn't the best place for that constraint and as evidenced by the original issue, it's having unintended effects when there are empty string elements in the table that wind up being evaluated (e.g. they have stats that get evaluated, etc).

**Comment by kbendick** [comment]
When I run the failed test suite locally by itself(in this case the failing tests are in `org.apache.iceberg.spark.extensions.TestCopyOnWriteDelete`), it passes just fine. So I think that there's either a complication due to timeout or some other transient error or that some of the tests are stepping on each others temporary directories etc due to inheritance possibly.  As the command `./gradlew check -Pquick=true -x javadoc` does also occasionally fail for me locally, though it does sometime pass as well. Either way, I don't think the failures are related at all.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java (hunk 4 above)
Comment (rdblue) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: I'm fine adding the test here, but I'd rather not change the existing cases. Would adding a new record test this without affecting the existing ones?

Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: Yes unfortunately, adding a new record to be returned from this function would mean we'd have to change many of the existing tests, as many of them have a hardcoded assumption on 10 records.

I can add a new test in another (possibly new) file if we'd prefer. But in order to place the test in here to be run against what seems to be one of the more comprehensive test suites for hitting various evaluators in various scenarios (i.e. it tests against tables partitioned by the empty string record, tables not partitioned by the empty string record, unpartitioned tables), we'd have to either change an existing record or update the tests to be aware of 11 records.

I'll take another look to see if there's an existing test suite where a new test can be easily added to cover this scenario without affecting other tests.

Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: If I add the following test to this file, it does cover some of the codepaths (albeit with a somewhat funny SQL query). This will still go over the `truncateBinary` function when the input needs to be truncated to a length of zero, although this time it will be because the predicate literal having a length of zero instead of the input data having a row with a length of zero.

But the following test would throw prior to this patch (and it doesn't require us to touch any of the input data for the other test suites).

Alternatively, I could add a new test file entirely.

```java
  @Test
  public void testPartitionedByDataStartsWithEmptyStringFilter() {
    File location = buildPartitionedTable("partitioned_by_data", PARTITION_BY_DATA, "data_ident", "data");

    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
        "path", location.toString())
    );

    IcebergSource source = new IcebergSource();
    DataSourceReader reader = source.createReader(options);
    pushFilters(reader, new StringStartsWith("data", ""));

    Assert.assertEquals(10, reader.planInputPartitions().size());
  }
```

Comment (kbendick) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:532]: I added a test that does not change any of the exising test data or add a new file at all, in `TestInclusiveMetricsEvaluator`. https://github.com/apache/iceberg/pull/2081/files#diff-a107d871a1a5d5c4029451db68cf5497ddc9432d90619e263270cf72c3536c23R73

I personally prefer changing the input data on the larger end to end test that covers many portions (`TestFilteredScan`) as they truly cover the situation where the bug was originally reported occurring (and where it is likely to happen based on usages of `truncateBinary`) - filtered scans with empty row data.

But I've added another possible way to test the reported scenario -`TestInclusiveMetricsEvaluator`.

If you'd like me to add a standalone spark test or something, let me know. Or if you prefer the new test, let me know and I can revert the change to the input records of the other test suite.

Thanks for your review @rdblue 🙂

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @kbendick! Good catch.

**Comment by kbendick** [comment]
> Thanks, @kbendick! Good catch.

My pleasure!
